### PR TITLE
1129 make membrane tooltips dynamic

### LINKED ISSUE #1129: Make membrane tooltips dynamic
read the membrane properties from SimulationParameters when entering the editor

### LINKED ISSUE #1709: Make membrane tooltip numbers relative to the current membrane
to more easily communicate what the effects are to the player. I think this requires: https://github.com/Revolutionary-Games/Thrive/issues/1129 to be done first before this is then pretty easy to do.

Currently the tooltips show the absolute effects of the membranes, which also results in the double membrane just having a bunch of +0's as it has the default stats so it's effect is listed as nothing.

### PR DESCRIPTION
**Pull request by Holomanga**
Membrane tooltips are now dynamic, and the numbers shown are relative to the currently selected membrane.

Screenshot 2020-10-25 at 18 19 00
Screenshot 2020-10-25 at 18 18 55

Closes #1129 
Closes #1709

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
I think in the case of no change, you should put a + in there, so it would say `+0%` as otherwise it can be confused that it will *be* `0%` if you select the option.

**Comment by hhyyrylainen** [comment]
I didn't review this yet, but I *think* you might be missing this:
- the membrane type should be sent to the GUI on editor entry (and also when loading a save)

**Comment by Holomanga** [comment]
The screenshot is a bit old; I made it display "+0%" in this commit: https://github.com/Revolutionary-Games/Thrive/pull/1763/commits/51ddc949ad9b51a6ae2cb314c2b670d0a8f72b98

Currently it looks like this when there are 0s:
Screenshot 2020-10-25 at 19 07 57

**Comment by hhyyrylainen** [comment]
> The screenshot is a bit old; I made it display "+0%" in this commit: 51ddc94

I see, sorry I didn't notice that. As I said I didn't do a full review yet, just commenting some of my initial impressions.

**Review by Kasterisk — DISMISSED** [review]
Looks good functionality-wise, a few nitpicks below

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 1]
```diff
@@ -268,6 +268,86 @@ public void WriteOrganelleProcessList(List<ProcessSpeedInformation> processes)
         }
     }
 
+    public void WriteMembraneModifierList(MembraneType referenceMembrane, MembraneType membraneType)
+    {
+        string[] modifierNames =
+        {
+            "Mobility", "Osmoregulation Cost", "Resource Absorption Speed", "Health",
+            "Physical Resistance", "Toxin Resistance",
+        };
```
Comment (Kasterisk) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:277]: Also I'm wondering if the modifiers of a membrane type should be put inside an array in the membranes.json (and each modifier entry contains its display name and value, might need a MembraneModifier object to hold those?) so we could loop through from a list of MembraneModifiers in a membrane type instead. Not sure if that's necessary though

Comment (hhyyrylainen) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:277]: It seems like this list is here, but then there is a special case used in the foreach loop?
To me this list looks like it shouldn't even exist.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 2]
```diff
@@ -268,6 +268,86 @@ public void WriteOrganelleProcessList(List<ProcessSpeedInformation> processes)
...
+        var modifierInfoLabels = new Dictionary<string, ModifierInfoLabel>();
+        var modifierValues = new Dictionary<string, float>();
+        foreach (var modifier in modifierNames)
+        {
+            modifierInfoLabels.Add(modifier, GetModifierInfo(modifier));
+            switch (modifier)
+            {
+                case "Mobility":
+                    modifierValues.Add(modifier, membraneType.MovementFactor - referenceMembrane.MovementFactor);
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:286]: Instead just linearly have the code from these cases in the method.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 3]
```diff
@@ -268,6 +268,86 @@ public void WriteOrganelleProcessList(List<ProcessSpeedInformation> processes)
...
+                    + modifierValues[modifierLabel.Key].ToString("F0", CultureInfo.CurrentCulture);
+            }
+            else
+            {
+                modifierLabel.Value.ModifierValue = ((modifierValues[modifierLabel.Key] >= 0) ? "+" : string.Empty)
+                    + (modifierValues[modifierLabel.Key] * 100).ToString("F0", CultureInfo.CurrentCulture)
+                    + "%";
+            }
+
+            if (modifierValues[modifierLabel.Key] > 0)
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:336]: Or maybe create a helper method for adding a new property line to contain this logic in it.

**Review by icedjuro — APPROVED** [review]
Apart from what HH and Kasterisk requested, I have no concerns.

**Comment by Kasterisk** [comment]
@Holomanga Seeing the PR has been stuck for a while, I might try taking over this if you don't mind.

**Comment by RacerBG** [comment]
@Kasterisk Could you add a text, for example "CURRENT" below the name of the used membrane or something like that? I believe this will help the players to more easily recognize which membrane they are using before trying to switch.

**Comment by Kasterisk** [comment]
I think the membrane button being highlighted is enough to indicate that it's the currently selected one. Though maybe we could add that text inside the tooltip (below the title name) imo. I'll see what others think on discord.

**Comment by hhyyrylainen** [comment]
Didn't we remove the "selected" text from the organelle selection menu? I think the selected membrane should be communicated similarly to the currently selected organelle.

**Comment by RacerBG** [comment]
> I think the membrane button being highlighted is enough to indicate that it's the currently selected one. Though maybe we could add that text inside the tooltip (below the title name) imo. I'll see what others think on discord.

Yes, I mean in the tooltip. :)

**Comment by Kasterisk** [comment]
> Yes, I mean in the tooltip. :)

Oh oops sorry, I thought you meant under the button name, which is how it was in older versions.

> Didn't we remove the "selected" text from the organelle selection menu? I think the selected membrane should be communicated similarly to the currently selected organelle.

I suppose the text could be added to the organelle tooltips too. If everyone is alright with that, I'll open an issue to readd the selected text on the tooltips I guess (since it's out of scope of this PR). But I'm still leaning towards just having the highlighted button to indicate that it's currently selected.

@RacerBG Currently the selected membrane has all its stats show 0 value on the tooltip, so maybe that could be used to tell which one is the current membrane instead.

**Comment by RacerBG** [comment]
> @RacerBG Currently the selected membrane has all its modifier value set to zero on the tooltip, so maybe that could be used to tell which one is the current membrane instead.

I'm fine with that, also you said that the current membrane is highlighted so it can be left as it is. It's up to you.

**Comment by hhyyrylainen** [comment]
Does this actually have changes to translations? As far as I can tell, only the reference line comment are changed.
If that's the case, then I think the translation files should be reset to what they are in master, it's unnecessary to cause a bunch of changes due to them in this PR.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Other than the unnecessary locale file changes, this looks good.

**Review by hhyyrylainen — APPROVED** [review]

**Pull request merged**